### ISSUE TITLE
Interact with LLM in JSON

### ISSUE THREAD

**Issue by DonggeLiu** [google/oss-fuzz-gen#743]
Currently we use XML-tag, it would be nice to [use JSON](https://github.com/google/oss-fuzz-gen/pull/736#discussion_r1870449044) as that seems to be more common, e.g., in Tool API.

One potential issue is escaping special characters. We have seen cases where LLM tend to add redundant `\`s, which may have performance impact.

**Comment by nugget-cloud** [comment]
@DonggeLiu Hi can you assign this issue to me

**Comment by DonggeLiu** [comment]
Hi @nugget-cloud Thanks for willing to help.
#744 could be a good starting point for you : )

**Comment by nugget-cloud** [comment]
Hi, @DonggeLiu 
I am trying to understand the code changes I need to make to interact with the LLM in JSON 
- in the prompts directory, change the XML to JSON 
- make changes in the llm_toolkit/prompts_builder and prompts and output_parser.py to handle JSON output 

Please correct me if I am wrong here

**Comment by DonggeLiu** [comment]
Thanks @nugget-cloud!
I suggest breaking this task into stages, starting with a simpler change before moving to more complex ones.

## Stage 1: Converting XML prompts to JSON
The goal here is to improve clarity in our requests to the LLM by switching from XML to JSON, which is a more commonly used format. The core logic remains the same.

TODOs:
* This is essentially [#744](https://github.com/google/oss-fuzz-gen/pull/744).
* You can fix the conflict and may also need to check for places (e.g., in `agents/`) where prompts are constructed dynamically instead of using templates in `prompts/`.
* No changes are needed for `prompt_builder.py` or `prompts.py` since we still expect the LLM to return plain text responses.

We can stop here if Stage 2 turns out to be too complex—your help is greatly appreciated regardless!


## Stage 2: Structuring LLM Responses in JSON
This stage requires deeper modifications, ensuring the LLM directly returns structured JSON responses rather than plain text.

TODOs:
1. Use [GenerationConfig](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/control-generated-output) to enforce JSON responses. This involves defining a response_schema, which varies based on the expected output.
2. To simplify schema management, we will likely need to modify `prompt_builder.py` and `prompts.py`, potentially introducing a new class that extends existing text-based ones but is tailored for JSON-based responses.


If done correctly, we shouldn't need major changes to output_parser.py, as the responses should already be well-formatted. Minor adjustments, such as handling unquoted text, may be required.

Once this is complete, you will have made significant progress, and we can confidently close this issue.


## Stage 3 (Optional): Encapsulating LLM Interactions as [Function Calls](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/function-calling)
If you're interested in exploring this further, we could encapsulate LLM interactions as function calls. The idea is to explicitly define how the LLM's responses will be used by a function and request it to return structured parameter values accordingly.

For example, instead of directly asking the LLM to generate a fuzz target, we define a function that takes a fuzz target as a parameter and instruct the LLM to generate the appropriate parameter value.

In practice, this is more complex, but much of the groundwork has already been done in [#731](https://github.com/google/oss-fuzz-gen/pull/731), which I put on hold due to other priorities. Feel free to pick it up if you're interested!

-----
This is exploratory work, meaning it may improve the results or lead to lower performance. For example, due to issues in handling quoted/unquoted strings in JSON, worse LLM responses, or unexpected compatibility issues. Unfortunately, if the outcome turns out to be worse than the current approach (e.g., a lower fuzz target build rate), we won’t be able to merge the changes.

**Comment by moizulfiqar** [comment]
Hi @DonggeLiu! Is this issue resolved yet?
